### ISSUE TITLE
[Homework1] Fuzzing issue on hw1

### ISSUE THREAD

**Issue by junsooo** [kaist-cp/cs420#340]
When you're asking a question, please make sure:

- [x] You clarify your questions as clear as possible.  If possible, please make it a yes/no question.  If possible, please summarize your question in one sentence at the beginning of an issue.

--
Status: pass test_examples_write_c, not pass fuzz test with below problem.

Problem Description(On cs420 server):

command: ```python3 tests/fuzz.py --print -n10```
Result is:
```
Building KECC..
    Finished release [optimized] target(s) in 0.12s
Fuzzing with 10 test cases.
Test case #0 (skipped: 0)
Traceback (most recent call last):
  File "tests/fuzz.py", line 280, in <module>
    fuzz(tests_dir, fuzz_arg, args.num)
  File "tests/fuzz.py", line 238, in fuzz
    raise Exception("Test `{}` failed with exit code {}.".format(" ".join(args), proc.returncode))
Exception: Test `cargo run --features=build-bin --release --bin fuzz -- -p /home/s20******/kecc-public/tests/test_polished.c` failed with exit code 101.
```
At this time, i read some former issues like https://github.com/kaist-cp/cs420/issues/271.

tried command: ```python3 tests/fuzz.py -p -r```
Result:
```
Building KECC..
    Finished release [optimized] target(s) in 0.57s
===< 1181006 >===
running 4 interestingness tests in parallel
===< pass_unifdef :: 0 >===
===< pass_comments :: 0 >===
===< pass_ifs :: 0 >===
===< pass_includes :: 0 >===
===< pass_line_markers :: 0 >===
(10.8 %, 136681 bytes)
===< pass_blank :: 0 >===
(11.9 %, 134975 bytes)
===< pass_clang_binsrch :: replace-function-def-with-decl >===
(81.1 %, 28915 bytes)
===< pass_clang_binsrch :: remove-unused-function >===
(95.1 %, 7437 bytes)
===< pass_lines :: 0 >===
(95.0 %, 7643 bytes)
(100.0 %, 0 bytes)

The file being reduced has reached zero size; our work here is done.

If you did not want a zero size file, you must help C-Reduce out by
making sure that your interestingness test does not find files like
this to be interesting.


Traceback (most recent call last):
  File "tests/fuzz.py", line 278, in <module>
    creduce(tests_dir, fuzz_arg)
  File "tests/fuzz.py", line 200, in creduce
    raise Exception("Reducing test_reduced.c by `{}` failed with exit code {}.".format(" ".join(args), proc.returncode))
Exception: Reducing test_reduced.c by `creduce --tidy ./reduce-criteria.sh test_reduced.c` failed with exit code 1.
```
I thought no reduce content means test pass, so i tried submit my homework but it fails with this message on server.
```
Finished release [optimized] target(s) in 0.03s
Building KECC..
Fuzzing with 30 test cases.
Test case #0 (skipped: 0)
Traceback (most recent call last):
  File "tests/fuzz.py", line 280, in <module>
    fuzz(tests_dir, fuzz_arg, args.num)
  File "tests/fuzz.py", line 238, in fuzz
    raise Exception("Test `{}` failed with exit code {}.".format(" ".join(args), proc.returncode))
Exception: Test `cargo run --features=build-bin --release --bin fuzz -- -p /app/files/checks/3e5d4858-375e-484f-9958-1872293fd9f3/kecc-public/tests/test_polished.c` failed with exit code 101
```

At this time I watched https://github.com/kaist-cp/cs420/issues/170#issuecomment-637482746. And tried restoring the previous version of `tests/reduce-criteria-template.sh` but it didn't help. 

From now on, the environment is Mac m1 macbook.

Question 1:
I checked test_polished.c is made well. 
And with this command, `cargo run --features="build-bin" -- tests/test_polished.c -p`
I found parsing error. it was caused by test_polished.c's this line
```
float (float);
```
I think its not standard with c language. i don't know why this line exists. Is it problem?

Question 2:
I erased this Question 1's line and faced 
```
Finished dev [unoptimized + debuginfo] target(s) in 0.05s
Running `target/debug/kecc tests/test_polished.c -p`
thread 'main' panicked at 'assertion failed: struct_type.node.declarations.is_none()', src/c/parse.rs:323:33
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
``` 
I don't know why this error happens..

Maybe it could be m1 mac environment's problem..

Please answer these questions . thank you very much for your help

**Comment by minseongg** [comment]
> Problem Description(On cs420 server):
> 
> command: ```python3 tests/fuzz.py --print -n10```
> Result is:
> ```
> Building KECC..
>     Finished release [optimized] target(s) in 0.12s
> Fuzzing with 10 test cases.
> Test case #0 (skipped: 0)
> Traceback (most recent call last):
>   File "tests/fuzz.py", line 280, in <module>
>     fuzz(tests_dir, fuzz_arg, args.num)
>   File "tests/fuzz.py", line 238, in fuzz
>     raise Exception("Test `{}` failed with exit code {}.".format(" ".join(args), proc.returncode))
> Exception: Test `cargo run --features=build-bin --release --bin fuzz -- -p /home/s20******/kecc-public/tests/test_polished.c` failed with exit code 101.
> ```

At this time, what is the result of `cargo run --features=build-bin --bin fuzz -- -p tests/test_polished.c` on cs420 server?

> Question 1:
> I checked test_polished.c is made well. 
> And with this command, `cargo run --features="build-bin" -- tests/test_polished.c -p`
> I found parsing error. it was caused by test_polished.c's this line
> ```
> float (float);
> ```
> I think its not standard with c language. i don't know why this line exists. Is it problem?
> 
> Question 2:
> I erased this Question 1's line and faced 
> ```
> Finished dev [unoptimized + debuginfo] target(s) in 0.05s
> Running `target/debug/kecc tests/test_polished.c -p`
> thread 'main' panicked at 'assertion failed: struct_type.node.declarations.is_none()', src/c/parse.rs:323:33
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ``` 
> I don't know why this error happens..

Could you send an email to minseong.jang@kaist.ac.kr with your `tests/test_polished.c` and `tests/test.c` when the `tests/test_polished.c` was generated?

**Comment by junsooo** [comment]
> At this time, what is the result of cargo run --features=build-bin --bin fuzz -- -p tests/test_polished.c on cs420 server? If it fails with assertion failed, then a reduced buggy program should be generated by python3 tests/fuzz.py -p -r.
I finally found bug on my program and fixed, so this shouldn't be problem now.
I debugged with `cargo run --features="build-bin" -- tests/test_polished.c -p`

But the Question 1 and 2 could be problem(or some kind of rust or csmith's bug maybe? or m1 problem) so i send email containing test_polished.c and test.c to @minseongg.

**Comment by minseongg** [comment]
Question 1)

It was the problem of the fuzzer script that replacing `float fabsf(float);` into `float (float);`. ([tests/fuzz.py#L45](https://github.com/kaist-cp/kecc-public/blob/main/tests/fuzz.py#L45)) We will fix it soon.

Question 2)

That's because some elements coming from the header files in macOS in the preprocessing stage. For example, below is a code fragment in your `test_polished.c`.

L94-99

```c
# 237 "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/math.h" 3 4
 int __inline_signbitl(double __x) {
    struct { double __f; unsigned long long __u;} __u;
    __u.__f = __x;
    return (int)(__u.__u >> 63);
}
```

It declares `struct` within the function, which is an unsupported feature of KECC. ([src/c/parse.rs#L317-L323](https://github.com/kaist-cp/kecc-public/blob/main/src/c/parse.rs#L317-L323))

Unfortunately, we won't fix this because we only support Ubuntu 20.04 now. Sorry (https://github.com/kaist-cp/cs420/issues/165#issuecomment-637183214)

**Comment by junsooo** [comment]
Thank you for your kind explanation. I'm closing this issue

**Comment by minseongg** [comment]
> **Question 1**
> 
> It was the problem of the fuzzer script that replacing `float fabsf(float);` into `float (float);`. ([tests/fuzz.py#L45](https://github.com/kaist-cp/kecc-public/blob/main/tests/fuzz.py#L45)) We will fix it soon.

Fixed in [`68e6279`](https://github.com/kaist-cp/kecc-public/commit/667d892ee02b84874505eae713ef46ebb1ca8afc).
